Simplenote/TagListViewController.m

```diff
@@ -182,7 +182,7 @@ - (void)loadTags
 - (Tag *)tagWithName:(NSString *)tagName
 {
     for (Tag *tag in self.tagArray) {
-        if ([tag.name isEqualToString:tagName]) {
+        if ([tag.name compare:tagName options:NSCaseInsensitiveSearch] == NSOrderedSame) {
```

Comment:
this should be comparing the tag hashes and not raw textual comparisons because it has to normalize Unicode variations and also casing is locale-dependent and we want to make sure that our tag name collisions aren't dependent on which language is set for a given device

```objectivec
- (Tag *)tagWithName:(NSString *)tagName
{
	NSString *tagHash = [tagName asTagHash];
	for (Tag *tag in self.tagArray) {
		if ([tag.name.asTagHash isEqualToString:tagHash]) {
			…
```

Reply:
Tags themselves come from the backend 100% (and AFAIK aren't really localized on the device, thus, independent from the device language). 

Considering that, I'm not sure if we **need to** compare encoded tags (since the encoded string is generated starting off with the tagname we're checking already). Am I missing something? (and thank you for stepping by D!!)

Reply:
> aren't really localized on the device, thus, independent from the device language

the tag names aren't localized but the string comparison function _is_ and for non-English languages might return different results when comparing two strings than it would have if the device were set to English because casing rules are locale-dependent

> Considering that, I'm not sure if we need to compare encoded tags

if we want to properly differentiate tags or tag names then we have to compare the hash because there are several possible and likely ways that two tag names could be equivalent based on our rules: differences in Unicode representation as well as casing.

we might have _duplicate_ tag entries in the `tag` bucket and so imagine three tag entities erroneously exist with `name` values of the following: (copy/paste is transforming the characters so I've added the code points)
 - `ṩ` (U+0073 U+0323 U+0307) URL-encoding `s%CC%A3%CC%87`
 - `ṩ` (U+0073 U+0307 U+0323) URL-encoding `s%CC%87%CC%A3`
 - `ṩ` (U+1E69) URL-encoding `%E1%B9%A9`

these are all different characters and will compare differently but only one is the right tag, the one with the `%E1%B9%A9` hash. in this example we can see how performing a scan for `name` matches can propagate the errors forward in the "fixed" applications because the overlook the hash. it could very well be that the property tag entity is missing due to previously-erroneous renames. in such a case we can consider a corrective measure to try and find a tag with a matching name and then create a new entity for it, but that does introduce the same confusing complexity that lookup was offering

Reply:
Actually, [isEqualToString](https://developer.apple.com/documentation/foundation/nsstring/1407803-isequaltostring) is not localized. 

Localized comparison API(s) make it explicit in their names (for instance [localizedCompare:](https://developer.apple.com/documentation/foundation/nsstring/1416999-localizedcompare)).

```
The comparison uses the canonical representation of strings, which for a particular string is the length of the string plus the UTF-16 code units that make up the string. When this method compares two strings, if the individual Unicodes are the same, then the strings are equal, regardless of the backing store.
```

I'll try to build a collision in a unit test, nevertheless!

Reply:
Update: I've added this to a unit test, for posterity.

Indeed, **NSString** behaves **exactly** as you've outlined. However, Swift's equality does handle all the things. Thank you Dennis!!